Show "(guest)" by guest users' names, when setting enabled

chrisbobbe opened the pull request:
Stacked on my current revision for #5806.

I've put "(guest)" in italics where it's rendered with React components, since it was possible and convenient to do it there.

### DM compose content input placeholder text before / after:

| Before | After |
| --- | --- |
| <img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/3a1ba9e6-5440-4631-bab5-695ad3e03f07"> | <img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/2ef95fbd-9eae-4f2a-b354-daf39d5935fb"> |

### Other "after" screenshots:

<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/7ee8808c-5fb3-4576-abf8-35eeb4668d69">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/0a3a6520-f242-4fec-a1ef-901799f08371">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/e644a427-2449-40b2-b8c4-64aa18d4c4fc">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/53c4ea54-bb28-4131-b8bf-4c960d024673">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/4c7951f6-2bbd-4f0b-b18f-37fdb8f4fe81">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/52ba27e6-013e-4f48-b723-41eaafcc606f">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/aa01cdc8-4563-4859-bbc4-d9898a4a963f">
<img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/89f76ab0-0f3d-4d89-86fa-404884f5d566">


Fixes: #5804

chrisbobbe commented:
Here's a before / after for this commit. With, in both, an uncommitted patch to declare the `user_list_incomplete` client capability (see #5808):

CustomProfileFields: Use styled UserItem for (unknown user)

| Before | After |
| --- | --- |
| <img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/afd4520b-4f60-4094-a65a-bd988859ac37"> | <img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/beeab409-1a6d-4d5b-a201-20373b0663c0"> |

chrisbobbe commented:
Oh, and a whole-PR before-and-after showing the changes to `AvatarItem`, i.e., the avatar with an overlaid slashed-circle "remove" button, that shows you which users you've selected in the new-group-DM form. In particular this shows the effect of this commit:

AvatarItem: Stop carelessly amputating parts of people's names

| Before | After |
| --- | --- |
| <img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/f4577db4-4398-4ba8-bfa1-da9e455c42ca"> | <img width="552" alt="image" src="https://github.com/zulip/zulip-mobile/assets/22248748/126e914c-9b00-427e-a6cc-025f4ffdfaea"> |

chrisbobbe commented:
Thanks for the review! Revision pushed.

chrisbobbe commented:
Oops! Pushed to the wrong PR branch; back now.

gnprice reviewed (commented):
Thanks @chrisbobbe for building this! I agree with the strategy of starting by centralizing how we compute the name text. This looks to be quite thorough.

Several comments below, all of them small.

chrisbobbe commented on src/api/modelTypes.js:
```diff
@@ -311,8 +303,7 @@ export type User = {|
 
-  // TODO(server-4.0): New in FL 59
-  +role?: Role,
+  +role: Role,
```
Ah OK, makes sense. Thanks for the feedback!

chrisbobbe commented on src/compose/ComposeBox.js:
```diff
@@ -713,7 +714,14 @@ const ComposeBox: React$AbstractComponent<Props, ImperativeHandle> = forwardRef(
     return <AnnouncementOnly />;
   }
 
-  const placeholder = getComposeInputPlaceholder(narrow, ownUserId, allUsersById, streamsById);
+  const placeholder = getComposeInputPlaceholder(
+    narrow,
+    ownUserId,
+    allUsersById,
+    streamsById,
+    enableGuestUserIndicator,
+    _,
+  );
```
Yeah.

chrisbobbe commented:
Thanks for the review! Revision pushed, and I fixed the failing tests. I also added a TODO in a commit at the end for a place we don't yet have "(guest)" indicators: @-mentions in message content. Hmm.

gnprice commented:
Thanks! All LGTM now, except there are some rebase conflicts to resolve.

chrisbobbe commented:
Thanks! Revision pushed, resolving those conflicts.

gnprice commented on src/storage/__tests__/migrations-test.js:
```diff
@@ -102,13 +102,18 @@ describe('migrations', () => {
-  // What `base` becomes after all migrations.
-  const endBase = {
+  // What `base` becomes after migrations up through 62.
+  const base62 = {
     ...base52,
-    migrations: { version: 62 },
```
This `migrations` line should stay — it makes the description of this "base" value accurate, and it's what we do with the other "baseN" values.

gnprice reviewed (commented):
Thanks! Looks good except one small glitch below. Please go ahead and merge after that.

chrisbobbe commented:
Thanks! Done.
